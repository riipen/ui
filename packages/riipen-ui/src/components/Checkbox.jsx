import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import ThemeContext from "../styles/ThemeContext";
import withClasses from "../utils/withClasses";

import Typography from "./Typography";

class Checkbox extends React.Component {
  static propTypes = {
    /**
     * If `true`, the component is checked.
     */
    checked: PropTypes.bool,

    /**
     * An array of custom CSS classes to apply.
     */
    classes: PropTypes.array,

    /**
     * The color of the component. It supports those theme colors that make sense for this component.
     */
    color: PropTypes.oneOf(["primary", "secondary", "default"]),

    /**
     * If `true`, the checkbox will be disabled.
     */
    disabled: PropTypes.bool,

    /**
     * An error to display below the checkbox.
     */
    error: PropTypes.node,

    /**
     * Label text to display for the checkbox.
     */
    label: PropTypes.node,

    /**
     * If true, an asterisk will be appended to the end of the label.
     */
    required: PropTypes.bool,

    /**
     * A warning to display below the checkbox.
     */
    warning: PropTypes.node
  };

  static defaultProps = {
    checked: false,
    disabled: false,
    required: false
  };

  static contextType = ThemeContext;

  render() {
    const {
      checked,
      classes,
      color,
      disabled,
      error,
      label,
      required,
      warning,
      ...other
    } = this.props;

    const theme = this.context;

    const className = clsx("checkbox");

    return (
      <div className={clsx(classes)}>
        <label htmlFor={other.id}>
          <Typography>
            {label}
            {required && " *"}
          </Typography>
          <input
            checked={checked}
            className={className}
            disabled={disabled}
            type="checkbox"
            {...other}
          />
          <span
            className={clsx("checkmark", disabled ? "disabled" : null, color)}
          />
        </label>
        {error && (
          <Typography
            classes={[clsx("error")]}
            color="negative"
            variant="body2"
          >
            {error}
          </Typography>
        )}
        {warning && (
          <Typography
            classes={[clsx("error")]}
            color="secondary"
            variant="body2"
          >
            {warning}
          </Typography>
        )}
        <style jsx>{`
          div {
            min-height: 20px;
          }

          label {
            cursor: pointer;
            display: block;
            padding-left: 35px;
            position: relative;
            user-select: none;
          }

          /* Hide the browser's default checkbox */
          label input {
            cursor: pointer;
            height: 0;
            opacity: 0;
            position: absolute;
            width: 0;
          }

          /* Create a custom checkbox */
          .checkmark {
            background-color: ${theme.palette.common.white};
            border: 1px solid rgba(0, 0, 0, 0.23);
            border-radius: ${theme.shape.borderRadius.md};
            height: 20px;
            left: 0;
            position: absolute;
            top: 0;
            width: 20px;
          }

          label input:focus ~ .checkmark {
            outline: -webkit-focus-ring-color auto 5px;
          }

          /* When the checkbox is checked */
          label input:checked ~ .checkmark {
            background-color: ${theme.palette.grey[400]};
            border-color: ${theme.palette.grey[400]};
          }

          label input:checked ~ .primary {
            background-color: ${theme.palette.primary.main};
            border-color: ${theme.palette.primary.main};
          }
          label input:checked ~ .secondary {
            background-color: ${theme.palette.secondary.main};
            border-color: ${theme.palette.secondary.main};
          }

          /* Create the checkmark/indicator (hidden when not checked) */
          .checkmark:after {
            content: "";
            display: none;
            position: absolute;
          }

          /* Show the checkmark when checked */
          label input:checked ~ .checkmark:after {
            display: block;
          }

          /* Style the checkmark/indicator */
          label .checkmark:after {
            border: solid white;
            border-width: 0 3px 3px 0;
            height: 10px;
            left: 6px;
            top: 2px;
            transform: rotate(45deg);
            width: 5px;
          }

          label input:checked ~ .disabled {
            background-color: ${theme.palette.grey[300]};
            border-color: ${theme.palette.grey[300]};
            pointer-events: none;
          }

          label + :global(.error) {
            margin-left: 35px !important;
            margin-top: ${theme.spacing(2)}px;
          }
        `}</style>
      </div>
    );
  }
}

export default withClasses(Checkbox);
