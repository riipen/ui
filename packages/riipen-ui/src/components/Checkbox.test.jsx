import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import Checkbox from "./Checkbox";

describe("<Checkbox>", () => {
  it("renders without errors", () => {
    let error;

    try {
      mount(<Checkbox readOnly />);
    } catch (e) {
      error = e;
    }

    expect(error).toEqual(undefined);
  });

  it("renders correct snapshot", () => {
    const wrapper = mount(<Checkbox readOnly />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("default props", () => {
    it("sets correct default props", () => {
      const wrapper = mount(<Checkbox readOnly />);

      const component = wrapper.find("Checkbox");
      expect(component.props().checked).toEqual(false);
      expect(component.props().disabled).toEqual(false);
      expect(component.props().required).toEqual(false);
    });
  });

  describe("checked prop", () => {
    it("sets checked with given checked value", () => {
      const checked = true;

      const wrapper = mount(<Checkbox checked={checked} readOnly />);

      expect(
        wrapper
          .find("Checkbox")
          .find("input")
          .props().checked
      ).toEqual(checked);
    });
  });

  describe("classes prop", () => {
    it("applies custom classes", () => {
      const classes = ["classOne"];

      const wrapper = mount(<Checkbox classes={classes} readOnly />);

      expect(
        wrapper
          .find("Checkbox")
          .childAt(0)
          .hasClass(classes[0])
      ).toEqual(true);
    });
  });

  describe("color prop", () => {
    it("sets color class name with a valid color", () => {
      const color = "secondary";

      const wrapper = mount(<Checkbox color={color} readOnly />);

      expect(
        wrapper
          .find("Checkbox")
          .childAt(0)
          .childAt(0)
          .childAt(2)
          .hasClass(color)
      ).toEqual(true);
    });

    it("gives an error when given an invalid color", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<Checkbox color="pink" readOnly />);

      expect(errors).toHaveBeenCalledTimes(1);
    });
  });

  describe("disabled prop", () => {
    it("sets disabled with given disabled value", () => {
      const disabled = true;

      const wrapper = mount(<Checkbox disabled={disabled} readOnly />);

      expect(
        wrapper
          .find("Checkbox")
          .find("input")
          .props().disabled
      ).toEqual(disabled);
    });

    it("applies disabled class name when disabled is true", () => {
      const disabled = true;

      const wrapper = mount(<Checkbox disabled={disabled} readOnly />);

      expect(
        wrapper
          .find("Checkbox")
          .find(".checkmark")
          .hasClass("disabled")
      ).toEqual(true);
    });
  });

  describe("error prop", () => {
    it("displays given error", () => {
      const error = <span>error!</span>;

      const wrapper = mount(<Checkbox error={error} readOnly />);

      expect(wrapper.childAt(0).contains(error)).toEqual(true);
    });
  });

  describe("label prop", () => {
    it("displays given label in the label element", () => {
      const label = "this is a label";

      const wrapper = mount(<Checkbox label={label} readOnly />);

      expect(wrapper.find("label").contains(label)).toEqual(true);
    });
  });

  describe("required prop", () => {
    it("displays an asterisk at the end of the label when required is true", () => {
      const label = <span>this is a label</span>;
      const required = true;

      const wrapper = mount(
        <Checkbox required={required} label={label} readOnly />
      );

      expect(
        wrapper
          .find("Checkbox")
          .find("label")
          .text()
          .substr(-1)
      ).toEqual("*");
    });

    it("does not display an asterisk when required is false", () => {
      const label = "this is a label";
      const required = false;

      const wrapper = mount(
        <Checkbox required={required} label={label} readOnly />
      );

      expect(
        wrapper
          .find("Checkbox")
          .find("label")
          .text()
          .substr(-1)
      ).not.toEqual("*");
    });
  });

  describe("warning prop", () => {
    it("displays given warning", () => {
      const warning = <span>warning!</span>;

      const wrapper = mount(<Checkbox warning={warning} readOnly />);

      expect(wrapper.childAt(0).contains(warning)).toEqual(true);
    });
  });

  describe("focusVisible state", () => {
    it("applies focusVisible class when focus event occurs", () => {
      const wrapper = mount(<Checkbox readOnly />);

      wrapper.find("input").simulate("focus");

      expect(
        wrapper
          .find("Checkbox")
          .find("input")
          .hasClass("focusVisible")
      ).toEqual(true);
    });

    it("does not apply focusVisible class when blur event occurs", () => {
      const wrapper = mount(<Checkbox readOnly />);

      wrapper.find("input").simulate("focus");
      wrapper.find("input").simulate("blur");

      expect(
        wrapper
          .find("Checkbox")
          .find("input")
          .hasClass("focusVisible")
      ).toEqual(false);
    });
  });

  describe("onKeyDown event", () => {
    it("invokes handleKeyDown on valid keydown event", () => {
      const handler = jest.fn();

      const wrapper = mount(<Checkbox onKeyDown={handler} readOnly />);

      wrapper.find("input").simulate("keydown", { key: "Enter" });

      expect(handler).toHaveBeenCalledTimes(1);
    });
  });
});
