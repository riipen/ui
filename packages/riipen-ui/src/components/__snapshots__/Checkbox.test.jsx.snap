// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox> renders correct snapshot 1`] = `
<Component
  readOnly={true}
>
  <Checkbox
    checked={false}
    disabled={false}
    readOnly={true}
    required={false}
    theme={
      Object {
        "breakpoints": Object {
          "lg": 1380,
          "md": 960,
          "sm": 600,
          "xl": 1920,
          "xs": 0,
        },
        "palette": Object {
          "background": Object {
            "default": "#fafafa",
          },
          "common": Object {
            "black": "#000",
            "white": "#fff",
          },
          "disabled": "#cccbcb",
          "divider": "#ddd",
          "grey": Object {
            "100": "#f5f5f5",
            "200": "#eeeeee",
            "300": "#dddddd",
            "400": "#cccbcb",
            "50": "#fafafa",
            "500": "#9a9a9a",
            "600": "#747474",
            "800": "#424242",
            "900": "#212121",
            "A100": "#d5d5d5",
            "A400": "#373737",
          },
          "negative": Object {
            "contrast": "#fff",
            "dark": "#ba2222",
            "light": "#ea7e7e",
            "main": "#e95353",
          },
          "positive": Object {
            "contrast": "#fff",
            "dark": "#1b965c",
            "light": "#78dbac",
            "main": "#4cc88d",
          },
          "primary": Object {
            "contrast": "#fff",
            "dark": "#254364",
            "light": "#2c8ddc",
            "main": "#1f71ae",
          },
          "secondary": Object {
            "contrast": "#fff",
            "dark": "#ed7321",
            "light": "#f5bf95",
            "main": "#f58632",
          },
          "tertiary": Object {
            "contrast": "#fff",
            "dark": "#41978a",
            "light": "#78becc",
            "main": "#3caabb",
          },
          "text": Object {
            "disabled": "#cccbcb",
            "hint": "rgba(0, 0, 0, 0.38)",
            "primary": "#373737",
            "secondary": "#747474",
          },
          "warning": Object {
            "contrast": "#fff",
            "dark": "#cf890e",
            "light": "#ffd894",
            "main": "#febb46",
          },
        },
        "shadows": Object {
          "0": "none",
          "1": "0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12)",
          "10": "0px 6px 6px -3px rgba(0,0,0,0.2),0px 10px 14px 1px rgba(0,0,0,0.14),0px 4px 18px 3px rgba(0,0,0,0.12)",
          "11": "0px 6px 7px -4px rgba(0,0,0,0.2),0px 11px 15px 1px rgba(0,0,0,0.14),0px 4px 20px 3px rgba(0,0,0,0.12)",
          "12": "0px 7px 8px -4px rgba(0,0,0,0.2),0px 12px 17px 2px rgba(0,0,0,0.14),0px 5px 22px 4px rgba(0,0,0,0.12)",
          "13": "0px 7px 8px -4px rgba(0,0,0,0.2),0px 13px 19px 2px rgba(0,0,0,0.14),0px 5px 24px 4px rgba(0,0,0,0.12)",
          "14": "0px 7px 9px -4px rgba(0,0,0,0.2),0px 14px 21px 2px rgba(0,0,0,0.14),0px 5px 26px 4px rgba(0,0,0,0.12)",
          "15": "0px 8px 9px -5px rgba(0,0,0,0.2),0px 15px 22px 2px rgba(0,0,0,0.14),0px 6px 28px 5px rgba(0,0,0,0.12)",
          "16": "0px 8px 10px -5px rgba(0,0,0,0.2),0px 16px 24px 2px rgba(0,0,0,0.14),0px 6px 30px 5px rgba(0,0,0,0.12)",
          "17": "0px 8px 11px -5px rgba(0,0,0,0.2),0px 17px 26px 2px rgba(0,0,0,0.14),0px 6px 32px 5px rgba(0,0,0,0.12)",
          "18": "0px 9px 11px -5px rgba(0,0,0,0.2),0px 18px 28px 2px rgba(0,0,0,0.14),0px 7px 34px 6px rgba(0,0,0,0.12)",
          "19": "0px 9px 12px -6px rgba(0,0,0,0.2),0px 19px 29px 2px rgba(0,0,0,0.14),0px 7px 36px 6px rgba(0,0,0,0.12)",
          "2": "0px 3px 1px -2px rgba(0,0,0,0.2),0px 2px 2px 0px rgba(0,0,0,0.14),0px 1px 5px 0px rgba(0,0,0,0.12)",
          "20": "0px 10px 13px -6px rgba(0,0,0,0.2),0px 20px 31px 3px rgba(0,0,0,0.14),0px 8px 38px 7px rgba(0,0,0,0.12)",
          "21": "0px 10px 13px -6px rgba(0,0,0,0.2),0px 21px 33px 3px rgba(0,0,0,0.14),0px 8px 40px 7px rgba(0,0,0,0.12)",
          "22": "0px 10px 14px -6px rgba(0,0,0,0.2),0px 22px 35px 3px rgba(0,0,0,0.14),0px 8px 42px 7px rgba(0,0,0,0.12)",
          "23": "0px 11px 14px -7px rgba(0,0,0,0.2),0px 23px 36px 3px rgba(0,0,0,0.14),0px 9px 44px 8px rgba(0,0,0,0.12)",
          "24": "0px 11px 15px -7px rgba(0,0,0,0.2),0px 24px 38px 3px rgba(0,0,0,0.14),0px 9px 46px 8px rgba(0,0,0,0.12)",
          "3": "0px 3px 3px -2px rgba(0,0,0,0.2),0px 3px 4px 0px rgba(0,0,0,0.14),0px 1px 8px 0px rgba(0,0,0,0.12)",
          "4": "0px 2px 4px -1px rgba(0,0,0,0.2),0px 4px 5px 0px rgba(0,0,0,0.14),0px 1px 10px 0px rgba(0,0,0,0.12)",
          "5": "0px 3px 5px -1px rgba(0,0,0,0.2),0px 5px 8px 0px rgba(0,0,0,0.14),0px 1px 14px 0px rgba(0,0,0,0.12)",
          "6": "0px 3px 5px -1px rgba(0,0,0,0.2),0px 6px 10px 0px rgba(0,0,0,0.14),0px 1px 18px 0px rgba(0,0,0,0.12)",
          "7": "0px 4px 5px -2px rgba(0,0,0,0.2),0px 7px 10px 1px rgba(0,0,0,0.14),0px 2px 16px 1px rgba(0,0,0,0.12)",
          "8": "0px 5px 5px -3px rgba(0,0,0,0.2),0px 8px 10px 1px rgba(0,0,0,0.14),0px 3px 14px 2px rgba(0,0,0,0.12)",
          "9": "0px 5px 6px -3px rgba(0,0,0,0.2),0px 9px 12px 1px rgba(0,0,0,0.14),0px 3px 16px 2px rgba(0,0,0,0.12)",
        },
        "shape": Object {
          "borderRadius": Object {
            "lg": "6px",
            "md": "4px",
            "sm": "3px",
            "xl": "8px",
          },
        },
        "spacing": [Function],
        "transitions": Object {
          "create": [Function],
          "duration": Object {
            "complex": 375,
            "short": 250,
            "shorter": 200,
            "shortest": 150,
            "standard": 300,
          },
          "easing": Object {
            "easeIn": "cubic-bezier(0.4, 0, 1, 1)",
            "easeInOut": "cubic-bezier(0.4, 0, 0.2, 1)",
            "easeOut": "cubic-bezier(0.0, 0, 0.2, 1)",
            "sharp": "cubic-bezier(0.4, 0, 0.6, 1)",
          },
        },
        "typography": Object {
          "body1": Object {
            "fontFamily": "Roboto, Helvetica, Arial, sans-serif",
            "fontSize": "14px",
            "fontWeight": 400,
            "letterSpacing": "0.00938em",
            "lineHeight": "20px",
          },
          "body2": Object {
            "fontFamily": "Roboto, Helvetica, Arial, sans-serif",
            "fontSize": "12px",
            "fontWeight": 400,
            "letterSpacing": "0.01071em",
            "lineHeight": "17px",
          },
          "body3": Object {
            "fontFamily": "Roboto, Helvetica, Arial, sans-serif",
            "fontSize": "10px",
            "fontWeight": 400,
            "letterSpacing": "0.01071em",
            "lineHeight": "14px",
          },
          "button": Object {
            "fontFamily": "Roboto, Helvetica, Arial, sans-serif",
            "fontSize": "14px",
            "fontWeight": 500,
            "letterSpacing": "0.02857em",
            "lineHeight": 1.75,
          },
          "fontFamily": "Roboto, Helvetica, Arial, sans-serif",
          "fontWeight": Object {
            "bold": 700,
            "light": 300,
            "medium": 500,
            "regular": 400,
          },
          "h1": Object {
            "fontFamily": "Roboto, Helvetica, Arial, sans-serif",
            "fontSize": "40px",
            "fontWeight": 500,
            "letterSpacing": "-0.01562em",
            "lineHeight": "56px",
          },
          "h2": Object {
            "fontFamily": "Roboto, Helvetica, Arial, sans-serif",
            "fontSize": "30px",
            "fontWeight": 500,
            "letterSpacing": "-0.00833em",
            "lineHeight": "42px",
          },
          "h3": Object {
            "fontFamily": "Roboto, Helvetica, Arial, sans-serif",
            "fontSize": "22px",
            "fontWeight": 500,
            "letterSpacing": "0em",
            "lineHeight": "32px",
          },
          "h4": Object {
            "fontFamily": "Roboto, Helvetica, Arial, sans-serif",
            "fontSize": "20px",
            "fontWeight": 400,
            "letterSpacing": "-0.00735em",
            "lineHeight": "30px",
          },
          "h5": Object {
            "fontFamily": "Roboto, Helvetica, Arial, sans-serif",
            "fontSize": "16px",
            "fontWeight": 400,
            "letterSpacing": "0rem",
            "lineHeight": "23px",
          },
        },
        "zIndex": Object {
          "above": 6,
          "high": 4,
          "highest": 5,
          "low": 2,
          "lowest": 1,
          "middle": 3,
          "zero": 0,
        },
      }
    }
  >
    <div
      className="jsx-1775705211 "
    >
      <label
        className="jsx-1775705211"
      >
        <withClasses(Typography)>
          <Typography
            classes={
              Array [
                "riipen",
                "riipen-typography",
              ]
            }
            color="inherit"
            display="initial"
            gutter={false}
            textAlign="inherit"
            textTransform="inherit"
            variant="body1"
          >
            <p
              className="jsx-3132880061 root color-inherit initial body1 align-inherit transform-inherit riipen riipen-typography"
            />
            <JSXStyle
              dynamic={
                Array [
                  "#424242",
                  "Roboto, Helvetica, Arial, sans-serif",
                  400,
                  "14px",
                  "20px",
                  "0.00938em",
                  "#424242",
                  "Roboto, Helvetica, Arial, sans-serif",
                  400,
                  "12px",
                  "17px",
                  "0.01071em",
                  "#424242",
                  "Roboto, Helvetica, Arial, sans-serif",
                  400,
                  "10px",
                  "14px",
                  "0.01071em",
                  "#373737",
                  "Roboto, Helvetica, Arial, sans-serif",
                  500,
                  "40px",
                  "56px",
                  "-0.01562em",
                  "#373737",
                  "Roboto, Helvetica, Arial, sans-serif",
                  500,
                  "30px",
                  "42px",
                  "-0.00833em",
                  "#373737",
                  "Roboto, Helvetica, Arial, sans-serif",
                  500,
                  "22px",
                  "32px",
                  "0em",
                  "#373737",
                  "Roboto, Helvetica, Arial, sans-serif",
                  400,
                  "20px",
                  "30px",
                  "-0.00735em",
                  "#373737",
                  "Roboto, Helvetica, Arial, sans-serif",
                  400,
                  "16px",
                  "23px",
                  "0rem",
                  "#000",
                  "#747474",
                  "#424242",
                  "#373737",
                  "#1f71ae",
                  "#f58632",
                  "#3caabb",
                  "#4cc88d",
                  "#e95353",
                  "#febb46",
                  "#fff",
                  700,
                  300,
                  500,
                  400,
                ]
              }
              id="1786047853"
            />
          </Typography>
        </withClasses(Typography)>
        <input
          checked={false}
          className="jsx-1775705211 checkbox"
          disabled={false}
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          readOnly={true}
          type="checkbox"
        />
        <span
          className="jsx-1775705211 checkmark"
        />
      </label>
      <JSXStyle
        dynamic={
          Array [
            "#fff",
            "4px",
            "#cccbcb",
            "#cccbcb",
            "#1f71ae",
            "#1f71ae",
            "#f58632",
            "#f58632",
            "#dddddd",
            "#dddddd",
            10,
          ]
        }
        id="3340623778"
      />
    </div>
  </Checkbox>
</Component>
`;
